### DISCUSSION TITLE
Attribute Pay records to user using email address

### DISCUSSION THREAD

**Discussion by gczh** [pay-rails/pay#1111] [Q&A]
At the moment, we use payment links generated on Stripe for our pricing tables hosted on a landing page builder.

Customers create a new subscription by paying using those payment links (which is different for each Subscription plan)

We're using Pay gem which register the Stripe webhook events, but they don't attribute the subscription and other related data to the User in our app through their email address.

The email address is used on the payment, and on our app.

Is there anyway to sync Stripe data (subscription, payments, etc) with the email address without hacking around too much? Or maybe, is there a better flow I can use here?

**Comment by excid3** [comment]
You can add a custom webhook to create the Pay::Customer by email address and then sync the subscription.

**Reply by gczh** [reply]
Ahh, got it – so I'd have to use a custom webhook. Appreciate you pointing me in the right direction 🙏 

When a payment is made through a payment link, there are quite a number of webhook events. Which one should I be using? I'm assuming `invoice.paid` as that means the payment was successfully made for the subscription?

As for syncing the subscription, would that be something like the following?

```ruby
user = User.find_by(email: email)

# manually pulling the customer for now to test how I'd sync the stripe customer from webhook to a Pay Customer
stripe_customer = Stripe::Customer.list(email: user.email, limit: 1).first

# Create a pay customer
Pay::Customer.create(
  owner: user,
  processor: :stripe,
  processor_id: stripe_customer.id
)

# Sync the subscription
user.payment_processor.sync_subscriptions
```

I tried the above but when I do `user.payment_processor.subscriptions`, I get an empty array.

Also, would there be an issue if say, a user has already created an account with the email address?

**Reply by excid3** [reply]
To create the Pay::Customer, you actually want to use:

```ruby
user.set_payment_processor(:stripe, processor_id: stripe_customer.id)
```

We use STI now for the Pay::Customer model, so without setting the `type` column, it may not work right.

An easier option might be to create the Stripe Customer with Pay and then embed the pricing table including their session: https://docs.stripe.com/payments/checkout/pricing-table#customer-session

Then I think the webhooks will take care of the rest.

**Reply by gczh** [reply]
> To create the Pay::Customer, you actually want to use:
> 
> ```ruby
> user.set_payment_processor(:stripe, processor_id: stripe_customer.id)
> ```
> 
> We use STI now for the Pay::Customer model, so without setting the `type` column, it may not work right.
> 
> An easier option might be to create the Stripe Customer with Pay and then embed the pricing table including their session: https://docs.stripe.com/payments/checkout/pricing-table#customer-session
> 
> Then I think the webhooks will take care of the rest.

Thanks @excid3, I just did the following manually and it works! 

```ruby
# TODO: shift the code into webhooks processing
user = Account.find_by(email: email)
stripe_customer = Stripe::Customer.list(email: user.email, limit: 1).first

user.set_payment_processor(:stripe, processor_id: stripe_customer.id)
user.payment_processor.sync_subscriptions
user.payment_processor.subscribed? # true
```

Your recommendation did the trick!

> ```ruby
> user.set_payment_processor(:stripe, processor_id: stripe_customer.id)
> ```

Now I just need to setup the custom webhook for that. I'm thinking of `stripe.customer.created` like the following:

```ruby
# app/webhooks/sync_stripe_customer.rb
class SyncStripeCustomer
  def call (event)
    stripe_customer_id = event.data.object["id"]
    email = event.data.object["email"]

    return if stripe_customer_id.blank? || email.blank?

    account = Account.find_by(email: email)
    return if account.blank?

    account.set_payment_processor(:stripe, processor_id: stripe_customer_id)
    account.payment_processor.sync_subscriptions
  end
end

# config/initializers/pay.rb
ActiveSupport.on_load(:pay) do
  Pay::Webhooks.delegator.subscribe "stripe.customer.created", SyncStripeCustomer.new
end
```

## A few questions though...

**Would adding `SyncStripeCustomer` conflict with what Pay is already doing to process the webhook events?** 

I'm not too sure if it would cause issues:

1. If the customer has already created an account with us with the same email, but not yet made any subscriptions
2. The above, but the customer might have added a payment method via `billing_portal`, and then adds a subscription via the payment link we linked to from our landing page instead
